feat: extend debian support up to bookworm

deubert-it opened the pull request:
see #243

deubert-it commented:
Regarding the failing checks/tests: I don't think the fails are related to the provided changes.

I also glanced over the spec files and did not see any debian codename related tests that might cover the provided changes, so it should generally be safe to merge imho.

smortex commented:
Hey :wave:!

The failures you see are caused by Debian 9 being EOL (so the tooling to test on this OS was removed) and a bug in Puppet 7 for which we have a workaround in the CI code… when this CI code is up-to-date.  Unfortunately, this module has been lagging behind for quite some time, #238 is about updating the module to update the CI code, but as you can see the module code quality does not match our expectations ATM.  So it's unfortunately hardly possible to work "correctly" on this module until #238 is merged…

If you want to help and have some time to offer, you are welcome to add commits on top of the branch of this PR and submit fixes in a new PR (luckily the auto-correct feature of rubocop will fix most issues, but be prepared to see new failures when you fix issues that prevent the next tests to be run). If you want to go this route, please note you can join #voxpupuli on IRC for assistance and guidance :grinning: 

When the CI code is up-to-date, updating `metadata.json` to add the Debian releases will make the CI run on these systems and ensure no regression is introduced on them.

smortex commented:
I gave it a shot. Can you please rebase your changes on top of the branch of #245 and add the versions in `metadata.json`?

deubert-it commented:
Hi @smortex,

thanks a lot for your help. I'm not sure the rebase is as you wanted it, but probably also fixable.
I updated the metadata as advised and it looks like debian 10 and 11 are fine, 12 seems to not be tested.
I tested with ubuntu 22 but it doesn't seem to work, so I removed it again from the supported versions list.

I can not say anything about the failing archlinux test, but my code change 100% did not touch anything that relates to archlinux.

Regarding debian 12 and ubuntu 22, from upgrading other systems myself, I know that it is required to now use gpg for apt sources and specify signing keys in the apt repo configuration ("[signed-by ...]"). (see https://wiki.debian.org/DebianRepository/UseThirdParty)
Do you know if the puppet apt_repo resource reflects that already or will we need to provide a solution in every module that wants to support debian 12/ubuntu 22?

However, I'm interested in getting the debian 11 support released, so we should probably move the debian 12/ubuntu 22 upgrade to a separate issue.

smortex commented:
This looks good! The ArchLinux failure is currently a known-issue and you can ignore it. Debian 12 was released a few days ago, and we need Puppet packages that do not exist yet to tests it thoroughly, so for now it is not part of the test matrix.  When available, adding it should be straightforward.

Regarding apt, it is something I still not have take the time to look at, and IMHO it should be managed at the apt module level for system-wide consistency.  I guess it would mean that the module should ensure the key does not exist in the legacy configuration and do exist in the new one. Might not be trivial, but if you work on it feel free to mention me so that I can help.

smortex commented:
Ah, just saw the branch was not rebased, so the commits are different.

Assuming you are in the branch `feature/243-debian-support` of your fork and you have a remote `origin` for `voxpupuli/puppet-gluster`:

```sh-session
$ git fetch --all           # Update what git thinks of the remotes
$ git rebase origin/debian9 # Move the commits of the branch on top of origin/debian9
$ git push -f               # -f required: we rewrote history
```

That way, when a branch is merged and some PR include its commits, it is automatically closed by GitHub which makes our lives easier and also helps when generating ChangeLogs.

deubert-it commented:
Hi @smortex , I (hope I) rebased the branch - is there anything else I can do to get this merged/released?

smortex commented:
Ooops! I merged some PRs which caused a bit of confusion :confused:.

FYI, I just rebased your branch on top of `master` and force-pushed to fix it!

smortex reviewed (approved):
LGTM!
